File: hydra-haskell/src/main/haskell/Hydra/Dsl/Graph.hs
module Hydra.Dsl.Graph where

import Hydra.Kernel
import Hydra.Dsl.Meta.Phantoms

import qualified Data.Map as M


comparisonLessThan :: TTerm Comparison
comparisonLessThan = injectUnit _Comparison _Comparison_lessThan

comparisonEqualTo :: TTerm Comparison
comparisonEqualTo = injectUnit _Comparison _Comparison_equalTo

comparisonGreaterThan :: TTerm Comparison
comparisonGreaterThan = injectUnit _Comparison _Comparison_greaterThan

graph :: TTerm (M.Map Name Binding)
    -> TTerm (M.Map Name (Maybe Term))
    -> TTerm (M.Map Name TypeScheme)
    -> TTerm Term
    -> TTerm (M.Map Name Primitive)
    -> TTerm (Maybe Graph)
    -> TTerm Graph
graph elements environment types body primitives schema = record _Graph [
    _Graph_elements>>: elements,
    _Graph_environment>>: environment,
    _Graph_types>>: types,
    _Graph_body>>: body,
    _Graph_primitives>>: primitives,
    _Graph_schema>>: schema]

graphElements :: TTerm Graph -> TTerm (M.Map Name Binding)
graphElements g = project _Graph _Graph_elements @@ g

graphEnvironment :: TTerm Graph -> TTerm (M.Map Name (Maybe Term))
graphEnvironment g = project _Graph _Graph_environment @@ g

graphTypes :: TTerm Graph -> TTerm (M.Map Name TypeScheme)
graphTypes g = project _Graph _Graph_types @@ g

graphBody :: TTerm Graph -> TTerm Term
graphBody g = project _Graph _Graph_body @@ g

graphPrimitives :: TTerm Graph -> TTerm (M.Map Name Primitive)
graphPrimitives g = project _Graph _Graph_primitives @@ g

graphSchema :: TTerm Graph -> TTerm (Maybe Graph)
graphSchema g = project _Graph _Graph_schema @@ g

graphWithElements :: TTerm Graph -> TTerm (M.Map Name Binding) -> TTerm Graph
graphWithElements g newElements = graph
    newElements
    (Hydra.Dsl.Graph.graphEnvironment g)
    (Hydra.Dsl.Graph.graphTypes g)
    (Hydra.Dsl.Graph.graphBody g)
    (Hydra.Dsl.Graph.graphPrimitives g)
    (Hydra.Dsl.Graph.graphSchema g)

graphWithEnvironment :: TTerm Graph -> TTerm (M.Map Name (Maybe Term)) -> TTerm Graph
graphWithEnvironment g newEnvironment = graph
    (Hydra.Dsl.Graph.graphElements g)
    newEnvironment
    (Hydra.Dsl.Graph.graphTypes g)
    (Hydra.Dsl.Graph.graphBody g)
    (Hydra.Dsl.Graph.graphPrimitives g)
    (Hydra.Dsl.Graph.graphSchema g)

graphWithTypes :: TTerm Graph -> TTerm (M.Map Name TypeScheme) -> TTerm Graph
graphWithTypes g newTypes = graph
    (Hydra.Dsl.Graph.graphElements g)
    (Hydra.Dsl.Graph.graphEnvironment g)
    newTypes
    (Hydra.Dsl.Graph.graphBody g)
    (Hydra.Dsl.Graph.graphPrimitives g)
    (Hydra.Dsl.Graph.graphSchema g)

graphWithBody :: TTerm Graph -> TTerm Term -> TTerm Graph
graphWithBody g newBody = graph
    (Hydra.Dsl.Graph.graphElements g)
    (Hydra.Dsl.Graph.graphEnvironment g)
    (Hydra.Dsl.Graph.graphTypes g)
    newBody
    (Hydra.Dsl.Graph.graphPrimitives g)
    (Hydra.Dsl.Graph.graphSchema g)

graphWithPrimitives :: TTerm Graph -> TTerm (M.Map Name Primitive) -> TTerm Graph
graphWithPrimitives g newPrimitives = graph
    (Hydra.Dsl.Graph.graphElements g)
    (Hydra.Dsl.Graph.graphEnvironment g)
    (Hydra.Dsl.Graph.graphTypes g)
    (Hydra.Dsl.Graph.graphBody g)
    newPrimitives
    (Hydra.Dsl.Graph.graphSchema g)

graphWithSchema :: TTerm Graph -> TTerm (Maybe Graph) -> TTerm Graph
graphWithSchema g newSchema = graph
    (Hydra.Dsl.Graph.graphElements g)
    (Hydra.Dsl.Graph.graphEnvironment g)
    (Hydra.Dsl.Graph.graphTypes g)
    (Hydra.Dsl.Graph.graphBody g)
    (Hydra.Dsl.Graph.graphPrimitives g)
    newSchema

primitive :: TTerm Name
    -> TTerm TypeScheme
    -> TTerm ([Term] -> Flow Graph Term)
    -> TTerm Primitive
primitive name typ implementation = record _Primitive [
    _Primitive_name>>: name,
    _Primitive_type>>: typ,
    _Primitive_implementation>>: implementation]

primitiveName :: TTerm Primitive -> TTerm Name
primitiveName p = project _Primitive _Primitive_name @@ p

primitiveType :: TTerm Primitive -> TTerm TypeScheme
primitiveType p = project _Primitive _Primitive_type @@ p

primitiveImplementation :: TTerm Primitive -> TTerm ([Term] -> Flow Graph Term)
primitiveImplementation p = project _Primitive _Primitive_implementation @@ p

primitiveWithType :: TTerm Primitive -> TTerm TypeScheme -> TTerm Primitive
primitiveWithType p newType = Hydra.Dsl.Graph.primitive
    (Hydra.Dsl.Graph.primitiveName p)
    newType
    (Hydra.Dsl.Graph.primitiveImplementation p)

typeClassEquality :: TTerm TypeClass
typeClassEquality = injectUnit _TypeClass _TypeClass_equality

typeClassOrdering :: TTerm TypeClass
typeClassOrdering = injectUnit _TypeClass _TypeClass_ordering

File: hydra-python/src/hydra/dsl/graph.py
"""Python implementations of hydra.dsl.graph primitives."""

import hydra.dsl.terms as terms
import hydra.gen.graph
from hydra.dsl.python import FrozenDict
from hydra.gen.core import Field, Name, Term, TypeScheme
from hydra.gen.graph import Element, Graph, Primitive
from hydra.gen.phantoms import TTerm


def element(
    name: TTerm[Name], term: TTerm[Term], mtyp: TTerm[TypeScheme | None]
) -> TTerm[Element]:
    """Create an element from a name, term, and optional type scheme."""
    return TTerm[Element](
        terms.record(
            hydra.gen.graph.ELEMENT__NAME,
            [
                Field(hydra.gen.graph.ELEMENT__NAME__NAME, name.value),
                Field(hydra.gen.graph.ELEMENT__TERM__NAME, term.value),
                Field(hydra.gen.graph.ELEMENT__TYPE__NAME, mtyp.value),
            ],
        )
    )


def element_name() -> TTerm[Name]:
    """Get the name of an element."""
    return TTerm[Name](
        terms.project(hydra.gen.graph.ELEMENT__NAME, hydra.gen.graph.ELEMENT__NAME__NAME)
    )


def graph(
    elements: TTerm[dict[Name, Element]],
    environment: TTerm[dict[Name, Term | None]],
    types: TTerm[dict[Name, TypeScheme]],
    body: TTerm[Term],
    primitives: TTerm[dict[Name, Primitive]],
    schema: TTerm[Graph | None],
) -> TTerm[Graph]:
    """Create a graph from a map of name/element bindings, a map of name/term bindings, a map of name/type scheme bindings, a body term, a map of name/primitive bindings, and an optional schema graph.

    Parameters
    ----------
    elements : TTerm[dict[Name, Element]]
        A map of name/element bindings.
    environment : TTerm[dict[Name, Term | None]]
        A map of name/term bindings.
    types : TTerm[dict[Name, TypeScheme]]
        A map of name/type scheme bindings.
    body : TTerm[Term]
        The body term.
    primitives : TTerm[dict[Name, Primitive]]
        A map of name/primitive bindings.
    schema : TTerm[Graph | None]
        An optional schema graph.

    Returns
    -------
    TTerm[Graph]
        A graph phantom term
    """
    return TTerm[Graph](
        terms.record(
            hydra.gen.graph.GRAPH__NAME,
            [
                Field(hydra.gen.graph.GRAPH__ELEMENTS__NAME, elements.value),
                Field(hydra.gen.graph.GRAPH__ENVIRONMENT__NAME, environment.value),
                Field(hydra.gen.graph.GRAPH__TYPES__NAME, types.value),
                Field(hydra.gen.graph.GRAPH__BODY__NAME, body.value),
                Field(hydra.gen.graph.GRAPH__PRIMITIVES__NAME, primitives.value),
                Field(hydra.gen.graph.GRAPH__SCHEMA__NAME, schema.value),
            ],
        )
    )


def graph_elements() -> TTerm[FrozenDict[Name, Element]]:
    """Get the elements of a graph."""
    return TTerm[FrozenDict[Name, Element]](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__ELEMENTS__NAME)
    )


def graph_environment() -> TTerm[FrozenDict[Name, Term | None]]:
    """Get the environment of a graph."""
    return TTerm[FrozenDict[Name, Term | None]](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__ENVIRONMENT__NAME)
    )


def graph_types() -> TTerm[FrozenDict[Name, TypeScheme]]:
    """Get the types of a graph."""
    return TTerm[FrozenDict[Name, TypeScheme]](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__TYPES__NAME)
    )


def graph_body() -> TTerm[Term]:
    """Get the body of a graph."""
    return TTerm[Term](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__BODY__NAME)
    )


def graph_primitives() -> TTerm[FrozenDict[Name, Primitive]]:
    """Get the primitives of a graph."""
    return TTerm[FrozenDict[Name, Primitive]](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__PRIMITIVES__NAME)
    )


def graph_schema() -> TTerm[Graph | None]:
    """Get the schema of a graph."""
    return TTerm[Graph | None](
        terms.project(hydra.gen.graph.GRAPH__NAME, hydra.gen.graph.GRAPH__SCHEMA__NAME)
    )


def primitive_name() -> TTerm[Name]:
    """Get the name of a primitive."""
    return TTerm[Name](
        terms.project(hydra.gen.graph.PRIMITIVE__NAME, hydra.gen.graph.PRIMITIVE__NAME__NAME)
    )


def primitive_type() -> TTerm[TypeScheme]:
    """Get the type of a primitive."""
    return TTerm[TypeScheme](
        terms.project(hydra.gen.graph.PRIMITIVE__NAME, hydra.gen.graph.PRIMITIVE__TYPE__NAME)
    )


# primitiveImplementation :: TTerm (Primitive -> ([Term] -> Flow Graph Term))
# primitiveImplementation = project _Primitive _Primitive_type
